src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -261,27 +261,41 @@ static const Type* bitshuffle_value(const TypeInteger* src_type, const TypeInteg
   }
 
   if (!mask_type->is_con()) {
-    int mask_max_bw;
-    int max_bw = bt == T_INT ? 32 : 64;
-    // Case 1) Mask value range includes -1.
-    if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
-      mask_max_bw = max_bw;
-    // Case 2) Mask value range is less than -1.
-    } else if (mask_type->hi_as_long() < -1L) {
-      mask_max_bw = max_bw - 1;
-    } else {
-    // Case 3) Mask value range only includes +ve values.
-      assert(mask_type->lo_as_long() >= 0, "");
-      jlong clz = count_leading_zeros(mask_type->hi_as_long());
-      clz = bt == T_INT ? clz - 32 : clz;
-      mask_max_bw = max_bw - clz;
-    }
     if ( opc == Op_CompressBits) {
-      lo = mask_max_bw == max_bw ? lo : 0L;
-      // Compress operation is inherently an unsigned operation and
-      // result value range is primarily dependent on true count
-      // of participating mask value.
-      hi = mask_max_bw < max_bw ? (1L << mask_max_bw) - 1 : src_type->hi_as_long();
+      // Pattern: Integer/Long.compress(src_type, mask_type)
+      int max_mask_bit_width;
```

Comment:
```suggestion
      int result_bit_width;
```
Is this bit width not about the result? It is really not about the mask.

Example:
`mask_type->hi_as_long() < -1L`

Here, the mask has the uppermost bit set, and so the bit width of it is the maximum 32 / 64 bits.

But we still can deduce that the result has one leading zero bit, and so the bit width of the result is either 31 or 63.

Reply:
Your interpretation is correct. But this is indeed the max_mask_bit_width, which constrains the upper bound of the result value.

Reply:
What I am saying is that calling it `max_mask_bit_width` seems **incorrect**. Your `Case 2` checks if the sign bit is set. In that case, `max_mask_bit_width` has to be `32` or `64`. But then you set it `max_mask_bit_width = mask_bit_width - 1;`. So now it would be `31` or `63`. That would be incorrect for the mask. But it would be correct for the result.